### ISSUE TITLE
False negatives with `__init__` and `NoReturn`.

### ISSUE THREAD

**Issue by saaketp** [microsoft/pyright#3188]
**Describe the bug**
The `__init__` method can be annotated with return type `NoReturn` but when the instance is actually constructed, pyright just ignores that annotation and assumes that the instance will be the same type as the class. This can cause false negatives.

**To Reproduce**
The following code type checks, but gives an error at runtime.

```python
# pyright: strict
from typing import overload, NoReturn
from typing_extensions import reveal_type


class A:
    @overload
    def __init__(self, a: None) -> NoReturn:
        pass

    @overload
    def __init__(self, a: int) -> None:
        pass

    def __init__(self, a: int | None) -> None:
        if a is None:
            raise ValueError
        self.a = a


a = A(None)  # Lines below this one are not marked unreachable by Pylance
reveal_type(a)  # Type of "a" is "A" (should have been NoReturn)
print(a.a)  # No errors here, it would have complained about Unknown type of a.a if a was NoReturn
```

**Expected behavior**
The `NoReturn` annotation should be honored, and the type of `A(None)` should be `NoReturn` and not `A` in the above example.

**VS Code extension or command-line**
Checked with both VSCode extension (Pylance v2022.3.2-pre.1) and command-line (pyright 1.1.229)

**Additional context**
mypy gives errors for the above code, but because of the wrong reasons, it just doesn't allow any other type than `None` for `__init__`.
I don't think that is the right behavior either, `__init__` can throw exceptions, so NoReturn should be allowed. (Looks like this is a known issue with mypy https://github.com/python/mypy/issues/10342).

**Comment by erictraut** [comment]
I think the current behavior is correct, or at least defensible. 

If `__init__` really is a `NoReturn`, then a call to the constructor would always throw an exception. Unless there is a metaclass with a custom `__call__` method that doesn't call through to the `__init__` method, there would be no way to construct the class. So I think this is an antipattern that type checkers should not need to deal with.

By contrast, mypy doesn't allow `NoReturn` as a valid return annotation for `__init__`.

**Comment by saaketp** [comment]
Okay, I was thinking of cases where one of the overloads would be `NoReturn` and the other overloads still allow to create the instance. It might be useful to say that one of the ways to construct the class could give an error.

But I realize that I could just remove the overload that could throw an exception by re-doing the overloads in a different way, so I didn't really need this.
